neonvm/controllers/virtualmachine_controller.go

```diff
@@ -412,33 +415,38 @@ func (r *VirtualMachineReconciler) doReconcile(ctx context.Context, virtualmachi
 					return err
 				}
 				// compare guest spec and count of plugged
-				if *virtualmachine.Spec.Guest.CPUs.Use > int32(len(cpusPlugged)) {
+				if int32(virtualmachine.Spec.Guest.CPUs.Use.Value()) > int32(len(cpusPlugged)) {
 					// going to plug one CPU
 					if err := QmpPlugCpu(virtualmachine); err != nil {
 						return err
 					}
-				} else if *virtualmachine.Spec.Guest.CPUs.Use < int32(len(cpusPlugged)) {
+				} else if int32(virtualmachine.Spec.Guest.CPUs.Use.Value()) < int32(len(cpusPlugged)) {
 					// going to unplug one CPU
 					if err := QmpUnplugCpu(virtualmachine); err != nil {
 						return err
 					}
 				}
+				// we should notify even if had'n plug/unplug anything because we may scale by cgroup
+				if err := notifyRunner(ctx, virtualmachine, *virtualmachine.Spec.Guest.CPUs.Use); err != nil {
+					return err
+				}
 			}
 
 			// get CPU details from QEMU and update status
-			cpusPlugged, _, err := QmpGetCpus(virtualmachine)
+			_, _, err := QmpGetCpus(virtualmachine)
 			if err != nil {
 				log.Error(err, "Failed to get CPU details from VirtualMachine", "VirtualMachine", virtualmachine.Name)
 				return err
 			}
-			if virtualmachine.Status.CPUs != len(cpusPlugged) {
+			// also get cgroup?
+			if virtualmachine.Status.CPUs.Cmp(*virtualmachine.Spec.Guest.CPUs.Use) != 0 {
 				// update status by count of CPU cores used in VM
-				virtualmachine.Status.CPUs = len(cpusPlugged)
+				virtualmachine.Status.CPUs = *virtualmachine.Spec.Guest.CPUs.Use
 				// record event about cpus used in VM
 				r.Recorder.Event(virtualmachine, "Normal", "CpuInfo",
-					fmt.Sprintf("VirtualMachine %s uses %d cpu cores",
+					fmt.Sprintf("VirtualMachine %s uses %v cpu cores",
 						virtualmachine.Name,
-						virtualmachine.Status.CPUs))
+						&virtualmachine.Status.CPUs))
```

Comment:
I was thinking of adding a GET endpoint too, because otherwise it just feels wrong to update the cgroup every second (every reconciliation loop iteration)

Reply:
BTW, how do we handle old runners that won't reply to this (and won't handle POST either)? We could have code that handles it gracefully, but I think we also have to work with full CPU for old runners in scheduler to avoid extra overcommit because of the absence of cgroup

Reply:
Oooh, yeah good point. I think a reasonable path forward is making "failed to connect" (or whatever equivalent network error) a warning, not hard error — for now. Then, once all old runners are gone, we change it to a hard error with a future update. HTTP error responses can still be a hard error though, because we should only be getting valid HTTP for new runners.

Reply:
We talked about this elsewhere; a different approach is now implemented (explicit versioning).